x-pack/plugins/ingest_pipelines/public/application/components/pipeline_editor/__jest__/processors/dot_expander.test.tsx

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { act } from 'react-dom/test-utils';
+import { setup, SetupResult, getProcessorValue } from './processor.helpers';
+
+const DOT_EXPANDER_TYPE = 'dot_expander';
+
+describe('Processor: Dot Expander', () => {
+  let onUpdate: jest.Mock;
+  let testBed: SetupResult;
+
+  beforeAll(() => {
+    jest.useFakeTimers();
+  });
+
+  afterAll(() => {
+    jest.useRealTimers();
+  });
+
+  beforeEach(async () => {
+    onUpdate = jest.fn();
+
+    await act(async () => {
+      testBed = await setup({
+        value: {
+          processors: [],
+        },
+        onFlyoutOpen: jest.fn(),
+        onUpdate,
+      });
+    });
+
+    testBed.component.update();
+
+    // Open flyout to add new processor
+    testBed.actions.addProcessor();
+    // Add type (the other fields are not visible until a type is selected)
+    await testBed.actions.addProcessorType(DOT_EXPANDER_TYPE);
+  });
+
+  test('prevents form submission if required fields are not provided', async () => {
+    const {
+      actions: { saveNewProcessor },
+      form,
+    } = testBed;
+
+    // Click submit button with only the type defined
+    await saveNewProcessor();
+
+    // Expect form error as "field" and "target_field" are required parameters
+    expect(form.getErrorsMessages()).toEqual(['A field value is required.']);
+  });
+
+  test('prevents form submission if field does not contain a . for the dot notation', async () => {
+    const {
+      actions: { saveNewProcessor },
+      form,
+      component,
+    } = testBed;
+
+    // Add "field" value (required)
+    form.setInputValue('fieldNameField.input', 'missing');
```

Comment:
As you are testing a validation to run and then display a message you need to wait for the validation to have ran. The validation run after a 500ms debounce time after the form input value has change.

As you correctly use the fake timer from jest the 500ms never pass and the validation does not run.

This is what you need to do

```js
// Wrap the `setInputValue` in an `act()` call as it modifies the internal value state of the field and mark it as "changed".
// This trigger the debounce which now waits... 
await act(async () => {
  form.setInputValue('fieldNameField.input', 'missing');
});
// Move ahead the debounce time which will then execute any validations
await act(async () => {
  jest.runAllTimers();
});
```

Reply:
@sebelga interestingly enough - we are not wrapping other instances of `form.setInputValue()` in `act()` (although I guess we probably should be?) or calling `runAllTimers()` in these processor tests. Is there something different about this specific validation?

On another note - should the `act()` be part of the actual `setInputValue` method in the testbed utils? I see it does call `component.update()` after the simulate change - [related code](https://github.com/elastic/kibana/blob/master/packages/kbn-test/src/jest/utils/testbed/testbed.ts#L196).

Reply:
Depending on what you are testing and if  there is code reacts to the field value change. You might not need to wrap with an `act` when changing the form input in certain cases. Here he is testing the result of the validation so he both needs the `act()` and wait for the debounce.

If for example you are testing that the correct payload is sent, then you don't need to wait for `useEffect` (that triggers the `validate()` on field). You can set the value of all fields synchronously and simulate the click of the "Send" button. That will call the `form.getData()` and all the values are there synchronously.

As you can see it in my latest test for preview, my helpers to modify field values all wrap with an `act` call (https://github.com/elastic/kibana/pull/106410/files#diff-8733b59b04f21f21eb6bf66bd48aba1334abad4ec93d8bde99ac1386e000f131R32)

> should the act() be part of the actual setInputValue method in the testbed utils?

It probably would have be good to do it there. Now it would require a lot of work to add it there and then remove it everywhere we have wrapped `setInputValue` with an `act` as it would be wrapped twice 😊 

Note that this is all due to hooks and the way state updates and react in `useEffect`. The `await act( ... )` makes sure all the lifecycle has ran.

Reply:
Thanks for the explanation @sebelga! 

I was thinking there were other test cases where we were verifying the input validation, but I see now most of the processor tests are only testing empty fields on submit, e.g.,

```
 test('prevents form submission if required fields are not provided', async () => {
    const {
      actions: { saveNewProcessor },
      form,
    } = testBed;

    // Click submit button with only the type defined
    await saveNewProcessor();

    // Expect form error as "field" and "target_field" are required parameters
    expect(form.getErrorsMessages()).toEqual(['A field value is required.']);
  });
```

Reply:
Yes and `saveNewProcessor` is wrapped

```js
async saveNewProcessor() {
  await act(async () => {
    find('addProcessorForm.submitButton').simulate('click');
  });
  component.update();
}
```

Reply:
I get it. I am not sure if this is spelled out explicitly in our CIT docs but I think it would be useful to do so. @sebelga @alisonelizabeth